prov/shm/src/smr_progress.c

```diff
@@ -891,7 +894,20 @@ static int smr_alloc_cmd_ctx(struct smr_ep *ep,
 	memcpy(&cmd_ctx->cmd, cmd, sizeof(*cmd));
 	cmd_ctx->ep = ep;
 
-	if (cmd->msg.hdr.op_src == smr_src_sar) {
+	if (cmd->msg.hdr.op_src == smr_src_inject) {
+		buf = ofi_buf_alloc(ep->unexp_buf_pool);
+		if (!buf) {
+			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+				"Error allocating buffer\n");
+			assert(0);
+		}
+		cmd_ctx->sar_entry = NULL;
+		slist_init(&cmd_ctx->buf_list);
+		slist_insert_tail(&buf->entry, &cmd_ctx->buf_list);
+		tx_buf = smr_get_ptr(ep->region, (size_t) cmd->msg.hdr.src_data);
+		memcpy(buf->buf, tx_buf->buf, cmd->msg.hdr.size);
```

Comment:
So we are using buffer of SAR_SIZE (32 K) for inject sized (4K) buffering? I think it may cause unnecessarily high memory usage when the number of unexp messages is high.

Reply:
Large (even 4k) unexpected message buffering if already going to case unnecessarily high memory usage. I did it this way to simplify the code. We could have two separate pools for 4k and 32k but that could also cause unnecessarily high memory usage since you can't mix and match (so if you've already allocated a bunch of 32k buffers but now you have a lot of 4k messages to buffer, you can't use that memory). The other solution would be to switch to 4k buffers all around and make the SAR messages take batch_size * 8 number of buffers but I think that would make the buffering code for SAR more complicated. Is there a method you would prefer?

Reply:
Let's go with the current simple approach.